### PR TITLE
[Core/MPI] allow to use a different data comm for partitioning

### PR DESCRIPTION
**Pull request by philbucher**
**Description**
See title.
Up to now we always used `MPI_COMM_WORLD`, now it is possible to also use other data comms

### CODE REVIEW

File: kratos/mpi/tests/test_distributed_import_model_part_utility.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        self.assertEqual(sub_smp.GetCommunicator().GlobalNumberOfConditions(), 1)
+
+        self.assertEqual(sub_smp.NumberOfSubModelParts(), 0)
+
+
+    def test_with_all_ranks(self):
+        self.__execute_test(in_memory=False, all_ranks=True)
+
+    def test_with_sub_comm(self):
+        self.skipTest("Not implemented")
```
Comment (philbucher) [inline kratos/mpi/tests/test_distributed_import_model_part_utility.py]: this test will come in a followup PR

**Comment by philbucher** [comment]
now also the `ParallelFillCommunicator` optionally accepts a `DataCommunicator`

I think this is the last change that is necessary from my side to make running solvers on a subset of MPI-processes possible

After this PR only minor adaptions in the apps should be necessary, I will start with StructuralMechanics and also add more tests there then

File: kratos/includes/fill_communicator.h [hunk 2]
```diff
@@ -90,6 +90,14 @@ class KRATOS_API(KRATOS_CORE) FillCommunicator
      */
     virtual void Execute();
 
+    /**
+     * @brief Execute the communicator fill
+     * This method is intended to perform the communicator filling In the current serial case it does nothing.
+     * For the parallel implementation see ParallelFillCommunicator.
+     * @param rModelPart DataCommunicator to be used for the partitioning
+     */
+    virtual void Execute(const DataCommunicator& rDataComm);
```
Comment (pooyan-dadvand) [inline kratos/includes/fill_communicator.h]: hmmm, I thought you would add this to the constructor. What would be the use case for changing it in Execute? Does it work if we call with a different one?

File: kratos/mpi/utilities/parallel_fill_communicator.cpp [hunk 3]
```diff
@@ -413,7 +418,7 @@ void ParallelFillCommunicator::InitializeParallelCommunicationMeshes(
     // Call the sub model part.
     for (ModelPart& r_sub_model_part : rModelPart.SubModelParts())
     {
-        ComputeCommunicationPlan(r_sub_model_part);
+        ComputeCommunicationPlan(r_sub_model_part, rModelPart.GetCommunicator().GetDataCommunicator());
```
Comment (pooyan-dadvand) [inline kratos/mpi/utilities/parallel_fill_communicator.cpp]: I feel that having some parts with the model part data communicator and execute with other data communicator is ambiguous and a source of obscure errors...

**Review by pooyan-dadvand — COMMENTED** [review]
I may miss the global context but I am not sure how safe is to have the parallel fill communicator with a different data communicator rather than the model part one. And even more, changing it in each call of the Execute

File: kratos/mpi/utilities/parallel_fill_communicator.cpp (hunk 3 above)
Comment (philbucher) [reply kratos/mpi/utilities/parallel_fill_communicator.cpp]: all of them have the same DataComm if you check the flow of the code
I can also pass it down directly but then I have to change a signature
in the end is the same

File: kratos/includes/fill_communicator.h (hunk 2 above)
Comment (philbucher) [reply kratos/includes/fill_communicator.h]: Works in either case, the one without the arg takes the default
I don't have a strong opinion, will change it

**Comment by philbucher** [comment]
> I may miss the global context but I am not sure how safe is to have the parallel fill communicator with a different data communicator rather than the model part one.

this is the point, so far the ParallelFillComm is the one assigning the (default) datacomm to the modelpart! => before calling the FillComm the ModelPart doesn't have a valid DataComm.
https://github.com/KratosMultiphysics/Kratos/blob/84e9c7128c638a1babdd248b7eb901fe35e66335/kratos/mpi/utilities/parallel_fill_communicator.cpp#L198

File: kratos/includes/fill_communicator.h (hunk 2 above)
Comment (philbucher) [reply kratos/includes/fill_communicator.h]: I think I did it to not have to save the DataComm as a member
In any case I will change it

Comment (philbucher) [reply kratos/includes/fill_communicator.h]: done, now it is clearer :)

File: kratos/mpi/utilities/parallel_fill_communicator.cpp (hunk 3 above)
Comment (philbucher) [reply kratos/mpi/utilities/parallel_fill_communicator.cpp]: improved :)

**Comment by philbucher** [comment]
@pooyan-dadvand I addressed all your comments, now I think the code looks better/cleaner

**Review by pooyan-dadvand — COMMENTED** [review]
@KratosMultiphysics/technical-committee approves the proposal and the change of the interface. The old FillCommunicator constructor will be deprecated in a separate PR.

**Review by roigcarlo — APPROVED** [review]
Ok!

**Pull request merged**